Tasks/DownloadPackageV1/extractor.ts

```diff
@@ -0,0 +1,125 @@
+
+export class Extractor {
+    public readonly win: boolean;
+
+    // 7zip
+    public xpSevenZipLocation: string;
+    public winSevenZipLocation: string = path.join(__dirname, "7zip/7z.exe");
+
+    private zipLocation: string;
+    private unzipLocation: string;
+
+    constructor(zipLocation: string, unzipLocation: string) {
+        this.zipLocation = zipLocation;
+        this.unzipLocation = unzipLocation;
+        this.win = os.platform() === "win32";
+        taskLib.debug("win: " + this.win);
+    }
+
+    private getSevenZipLocation(): string {
+        if (this.win) {
+            return this.winSevenZipLocation;
+        } else {
+            if (typeof this.xpSevenZipLocation === "undefined") {
+                this.xpSevenZipLocation = taskLib.which("7z", true);
+            }
+            return this.xpSevenZipLocation;
+        }
+    }
+
+    private isTar(file): boolean {
+        const name = file.toLowerCase();
+        // standard gnu-tar extension formats with recognized auto compression formats
+        // https://www.gnu.org/software/tar/manual/html_section/tar_69.html
+        return (
+            name.endsWith(".tar") || // no compression
+            name.endsWith(".tar.gz") || // gzip
+            name.endsWith(".tgz") || // gzip
+            name.endsWith(".taz") || // gzip
+            name.endsWith(".tar.z") || // compress
+            name.endsWith(".tar.bz2") || // bzip2
+            name.endsWith(".tz2") || // bzip2
+            name.endsWith(".tbz2") || // bzip2
+            name.endsWith(".tbz") || // bzip2
+            name.endsWith(".tar.lz") || // lzip
+            name.endsWith(".tar.lzma") || // lzma
+            name.endsWith(".tlz") || // lzma
+            name.endsWith(".tar.lzo") || // lzop
+            name.endsWith(".tar.xz") || // xz
+            name.endsWith(".txz")
+        ); // xz
+    }
+
+    private sevenZipExtract(file: string, destinationFolder: string) {
+        //We have to create our own 7Zip extract function as the vsts-task-tool-lib
+        //method uses 7zDec, which only decodes .7z archives
+        console.log(taskLib.loc("SevenZipExtractFile", file));
+        const sevenZip = taskLib.tool(this.getSevenZipLocation());
+        sevenZip.arg("x");
+        sevenZip.arg("-o" + destinationFolder);
+        sevenZip.arg(file);
+        const execResult = sevenZip.execSync();
+        if (execResult.code != taskLib.TaskResult.Succeeded) {
+            taskLib.debug("execResult: " + JSON.stringify(execResult));
+        }
+    }
+
+    async extractFile() {
+        const stats = taskLib.stats(this.zipLocation);
+        if (!stats) {
+            throw new Error(taskLib.loc("ExtractNonExistFile", this.zipLocation));
+        } else if (stats.isDirectory()) {
+            throw new Error(taskLib.loc("ExtractDirFailed", this.zipLocation));
+        }
+
+        const fileEnding = path.parse(this.zipLocation).ext;
+
+        if (this.win) {
+            if (".tar" === fileEnding) {
```

Comment:
This is unnecessarily complicated. There are 2 filetypes we're expecting to extract - nupkg and tgz. nupkg need to be unzipped, tgz need to be untarred. Instead of distributing 7z, consider an npm module like tar-stream.